// @carbon/charts/src/tests/test-environment.js
import * as Charts from "../index";
import { createChartHolder } from "./tools";
import { groupedBarData, groupedBarOptions } from "../../demo/demo-data";
export var data = groupedBarData;
export var options = Object.assign(groupedBarOptions, {
    title: "My chart"
});
var TestEnvironment = /** @class */ (function () {
    function TestEnvironment() {
        this.chartOptions = options;
        this.chartData = data;
    }
    TestEnvironment.prototype.render = function () {
        var holder = createChartHolder("scatter");
        this.chart = new Charts.ScatterChart(holder, {
            data: this.chartData,
            options: this.chartOptions
        });
    };
    TestEnvironment.prototype.destroy = function () {
        this.chart.destroy();
    };
    TestEnvironment.prototype.setChartOptions = function (func) {
        this.chartOptions = func(this.chartOptions);
    };
    TestEnvironment.prototype.getChartReference = function () {
        return this.chart;
    };
    return TestEnvironment;
}());
export { TestEnvironment };
